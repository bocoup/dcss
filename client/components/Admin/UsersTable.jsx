import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Pagination, Table } from '@components/UI';
import UserRows from './UserRows';

// import { getUsers } from '@actions/users';
const MOBILE_WIDTH = 767;
const ROWS_PER_PAGE = 10;
const USER_ROLES = {
  super_admin: 'Super Admin',
  admin: 'Admin',
  researcher: 'Researcher',
  facilitator: 'Facilitator',
  participant: 'Participant'
};

class UsersTable extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    const {
      activePage,
      cohort,
      columns,
      grantableRoles,
      onPageChange,
      rowsPerPage,
      pages,
      rows = {}
    } = this.props;

    const userRowsProps = { cohort, grantableRoles, rows, rowsPerPage };
    const tableRows = <UserRows {...userRowsProps} />;

    const IS_ON_MOBILE = window.innerWidth <= MOBILE_WIDTH;
    return (
      <Table striped selectable unstackable>
        {IS_ON_MOBILE ? (
          <Table.Header>
            <Table.Row>
              <Table.HeaderCell colSpan="7">
                <Pagination
                  borderless
                  name="users"
                  activePage={activePage}
                  siblingRange={1}
                  boundaryRange={0}
                  ellipsisItem={null}
                  firstItem={null}
                  lastItem={null}
                  onPageChange={onPageChange}
                  totalPages={pages}
                />
              </Table.HeaderCell>
            </Table.Row>
          </Table.Header>
        ) : null}
        <Table.Header>
          <Table.Row>
            {Object.entries(columns).map(([key, { className, content }]) => (
              <Table.HeaderCell key={key} className={className}>
                {content}
              </Table.HeaderCell>
            ))}
          </Table.Row>
        </Table.Header>
        <Table.Body className="ut__tbody">{tableRows}</Table.Body>
        <Table.Footer>
          <Table.Row>
            <Table.HeaderCell colSpan="7">
              <Pagination
                borderless
                name="users"
                activePage={activePage}
                siblingRange={1}
                boundaryRange={0}
                ellipsisItem={null}
                firstItem={null}
                lastItem={null}
                onPageChange={onPageChange}
                totalPages={pages}
              />
            </Table.HeaderCell>
          </Table.Row>
        </Table.Footer>
      </Table>
    );
  }
}

UsersTable.propTypes = {
  activePage: PropTypes.number,
  cohort: PropTypes.object,
  columns: PropTypes.object,
  grantableRoles: PropTypes.object,
  onPageChange: PropTypes.func,
  rowsPerPage: PropTypes.number,
  pages: PropTypes.number,
  rows: PropTypes.object
};

const mapStateToProps = (state, ownProps) => {
  let rowsPerPage = ROWS_PER_PAGE;
  let grantableRoles = USER_ROLES;

  if (ownProps.grantableRoles) {
    grantableRoles = ownProps.grantableRoles;
  }

  if (ownProps.rowsPerPage) {
    rowsPerPage = ownProps.rowsPerPage;
  }

  return { grantableRoles, rowsPerPage };
};

// const mapDispatchToProps = dispatch => ({
//   // getUsers: () => dispatch(getUsers())
// });

export default connect(
  mapStateToProps,
  null
)(UsersTable);
