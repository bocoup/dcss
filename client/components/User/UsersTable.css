
.users__col-xxlarge {
  width: 50%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap !important;
}

.users__col-xlarge {
  width: 40%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap !important;
}

.users__col-large {
  /*width: 30%;*/
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap !important;
}

.users__col-medium {
  width: 20%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap !important;
}

.users__col-medium-overflow-visible {
  width: 20%;
}

.users__col-small {
  width: 10%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap !important;
}

.ut__tbody {
  min-height: 231px;
}
