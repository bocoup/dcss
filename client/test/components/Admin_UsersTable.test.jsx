import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect
}));

import assert from 'assert';
import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import { mount, shallow } from 'enzyme';
import {
  fireEvent,
  prettyDOM,
  render,
  screen,
  waitFor
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import UsersTable from '../../components/Admin/UsersTable.jsx';

const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);

  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
});

afterEach(() => {
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
});

test('UsersTable', () => {
  expect(UsersTable).toBeDefined();
});

test('Render 1 1', async done => {
  const Component = UsersTable;

  const props = {
    ...commonProps,
    activePage: 0,
    cohort: {
      id: 1,
      created_at: '2020-08-31T14:01:08.656Z',
      name: 'A New Cohort That Exists Within Inline Props',
      runs: [
        {
          id: 11,
          user_id: 333,
          scenario_id: 99,
          created_at: '2020-03-28T19:44:03.069Z',
          updated_at: '2020-03-31T17:01:43.139Z',
          ended_at: '2020-03-31T17:01:43.128Z',
          consent_id: 8,
          consent_acknowledged_by_user: true,
          consent_granted_by_user: true,
          referrer_params: null,
          cohort_id: 1,
          run_id: 11
        }
      ],
      scenarios: [99],
      users: [
        {
          username: 'super',
          personalname: 'Super User',
          email: 'super@email.com',
          id: 999,
          roles: ['participant', 'super_admin'],
          is_anonymous: false,
          is_super: true
        },
        {
          username: 'facilitator',
          personalname: 'Facilitator User',
          email: 'facilitator@email.com',
          id: 555,
          roles: ['participant', 'facilitator', 'researcher'],
          is_anonymous: false,
          is_super: false
        },
        {
          username: 'researcher',
          personalname: 'Researcher User',
          email: 'researcher@email.com',
          id: 444,
          roles: ['participant', 'researcher'],
          is_anonymous: false,
          is_super: false
        },
        {
          username: 'participant',
          personalname: 'Participant User',
          email: 'participant@email.com',
          id: 333,
          roles: ['participant'],
          is_anonymous: false,
          is_super: false
        }
      ],
      roles: ['super', 'facilitator'],
      usersById: {
        999: {
          username: 'super',
          personalname: 'Super User',
          email: 'super@email.com',
          id: 999,
          roles: ['participant', 'super_admin'],
          is_anonymous: false,
          is_super: true
        },
        555: {
          username: 'facilitator',
          personalname: 'Facilitator User',
          email: 'facilitator@email.com',
          id: 555,
          roles: ['participant', 'facilitator', 'researcher'],
          is_anonymous: false,
          is_super: false
        },
        444: {
          username: 'researcher',
          personalname: 'Researcher User',
          email: 'researcher@email.com',
          id: 444,
          roles: ['participant', 'researcher'],
          is_anonymous: false,
          is_super: false
        },
        333: {
          username: 'participant',
          personalname: 'Participant User',
          email: 'participant@email.com',
          id: 333,
          roles: ['participant'],
          is_anonymous: false,
          is_super: false
        }
      }
    },
    columns: {},
    grantableRoles: {},
    onPageChange() {},
    rowsPerPage: 1,
    pages: 2,
    rows: {}
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  done();
});

/*{INJECTION}*/
