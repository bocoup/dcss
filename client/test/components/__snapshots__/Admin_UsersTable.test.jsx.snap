// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `
<DocumentFragment>
  <table
    class="ui selectable striped unstackable table"
  >
    <thead
      class=""
    >
      <tr
        class=""
      />
    </thead>
    <tbody
      class="ut__tbody"
    />
    <tfoot
      class=""
    >
      <tr
        class=""
      >
        <th
          class=""
          colspan="7"
        >
          <div
            aria-label="Pagination Navigation"
            class="ui borderless pagination menu"
            name="users"
            role="navigation"
          >
            <a
              aria-current="false"
              aria-disabled="false"
              aria-label="Previous item"
              class="item"
              tabindex="0"
              type="prevItem"
              value="1"
            >
              ⟨
            </a>
            <a
              aria-current="false"
              aria-disabled="false"
              class="item"
              tabindex="0"
              type="pageItem"
              value="1"
            >
              1
            </a>
            <a
              aria-current="false"
              aria-disabled="false"
              class="item"
              tabindex="0"
              type="pageItem"
              value="2"
            >
              2
            </a>
            <a
              aria-current="false"
              aria-disabled="false"
              aria-label="Next item"
              class="item"
              tabindex="0"
              type="nextItem"
              value="1"
            >
              ⟩
            </a>
          </div>
        </th>
      </tr>
    </tfoot>
  </table>
</DocumentFragment>
`;
